fdbserver/storageserver.actor.cpp

```diff
@@ -3212,36 +3208,101 @@ void StorageServerDisk::writeMutation( MutationRef mutation ) {
 		ASSERT(false);
 }
 
-void StorageServerDisk::writeMutations(const VectorRef<MutationRef>& mutations, Version debugVersion,
-                                       const char* debugContext) {
-	for (const auto& m : mutations) {
+ACTOR Future<int64_t> asyncWriteMutationsToCommitBuffer(StorageServer* data, VectorRef<MutationRef> mutations, Version debugVersion, const char* debugContext, int64_t ssCommitQuotaBytes) {
+	state int bytesWritten = 0;
+	state int i = 0;
+	for (;i < mutations.size(); i++) {
+		const auto& m = mutations[i];
 		DEBUG_MUTATION(debugContext, debugVersion, m).detail("UID", data->thisServerID);
 		if (m.type == MutationRef::SetValue) {
-			storage->set(KeyValueRef(m.param1, m.param2));
+			data->storage.set(KeyValueRef(m.param1, m.param2));
 		} else if (m.type == MutationRef::ClearRange) {
-			storage->clear(KeyRangeRef(m.param1, m.param2));
+			data->storage.clear(KeyRangeRef(m.param1, m.param2));
+		}
+		auto mutationBytes = mvccStorageBytes(m);
+		bytesWritten += mutationBytes;
+		ssCommitQuotaBytes -= mutationBytes;
+		if (data->storage.canPipelineCommits() && bytesWritten >= SERVER_KNOBS->STORAGE_COMMIT_PIPELINE_BYTES_PER_YIELD) {
+			bytesWritten = 0;
+			wait(yield());
 		}
 	}
+	return ssCommitQuotaBytes;
 }
 
-bool StorageServerDisk::makeVersionMutationsDurable( Version& prevStorageVersion, Version newStorageVersion, int64_t& bytesLeft ) {
-	if (bytesLeft <= 0) return true;
-
-	// Apply mutations from the mutationLog
-	auto u = data->getMutationLog().upper_bound(prevStorageVersion);
-	if (u != data->getMutationLog().end() && u->first <= newStorageVersion) {
-		VerUpdateRef const& v = u->second;
-		ASSERT( v.version > prevStorageVersion && v.version <= newStorageVersion );
-		// TODO(alexmiller): Update to version tracking.
-		DEBUG_KEY_RANGE("makeVersionMutationsDurable", v.version, KeyRangeRef());
-		writeMutations(v.mutations, v.version, "makeVersionDurable");
-		for (const auto& m : v.mutations) bytesLeft -= mvccStorageBytes(m);
-		prevStorageVersion = v.version;
-		return false;
-	} else {
-		prevStorageVersion = newStorageVersion;
-		return true;
+ACTOR Future<bool> asyncPrepareVersionsForCommit_impl(StorageServerDisk* self, StorageServer* data, Version desiredOldestVersion, Future<Void> durable, Future<Void>durableDelay) {
+	state int64_t ssCommitQuotaBytes = SERVER_KNOBS->STORAGE_COMMIT_BYTES;
+	state bool finalCommit = false;
+	state Version startOldestVersion = data->storageVersion();
+	state Version newOldestVersion = data->storageVersion();
+	state ActorCollection forgetter(true);
```

Comment:
Per offline discussion, convert these two vars to a SignalableActorCollection.

Reply:
Done.